### PR TITLE
[quant][fx] Support keyword arguments for functional linear

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79095

Summary:
Fixes: https://github.com/pytorch/pytorch/issues/78117
Fixes: https://github.com/pytorch/pytorch/issues/73463

This PR adds a normalization pass that normalizes all the args to keyword args in positional order and fixes lowering code that previously
only uses node.args to use both args and kwargs instead.

Also tried to add a test for F.conv2d, but since conv2d matches multiple schemas we are doing an extra schema match, and because we are using symbolic values
in `transform`, we don't have a schema match, so F.conv2d still fails with runtime errors. we can resolve this issue later when there is a need.

Another thing I'm considering is to do the normalization with real inputs instead of symbolic inputs and not rely on operator_schemas (which is based on torchscript),
and rely on inspect.signature, I tried this briefly but didn't get too far, it looks like we cannot get the python signature for `torch._C._nn.linear`, it might be possible to fix as well, but will need follow up discussions.

The goal for this PR is just to introduce normalization in our codebase so that we can adapt some downstream code to this, and also fix the F.linear issue.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_normalize_args

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D37163228](https://our.internmc.facebook.com/intern/diff/D37163228)

### CODE REVIEW

File: test/quantization/fx/test_quantize_fx.py [hunk 1]
```diff
@@ -7224,9 +7268,16 @@ def __init__(self):
                     self.emb = torch.nn.EmbeddingBag(num_embeddings=10, embedding_dim=12, mode='sum')
                     self.linear = torch.nn.Linear(12, 1).to(dtype=torch.float)
 
-                def forward(self, input: torch.Tensor, offsets: Optional[torch.Tensor] = None,
-                            per_sample_weights: Optional[torch.Tensor] = None):
-                    x = self.emb(input, offsets, per_sample_weights)
+                # Note: this does not work since the result is not scriptable
+                # because of these additional optional args
+                # def forward(self, input: torch.Tensor, offsets: Optional[torch.Tensor] = None,
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:7273]: Hi @supriyar, do you know how EmbeddingBag is used currently? do we pass in all optional arguments right now? and is scriptability still a requirement?

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 2]
```diff
@@ -25,14 +25,27 @@
...
-    ['running_mean', 'running_var', 'use_input_stats', 'momentum'],
+QOP_TO_ARG_INDEXES_TO_SKIP = {
+    # ['inplace']
+    torch._ops.ops.quantized.hardswish: [1],
+    # ['inplace']
+    torch._ops.ops.quantized.elu: [2],
+    # ['inplace'],
+    torch._ops.ops.quantized.dropout: [3],
+    # ['running_mean', 'running_var', 'use_input_stats', 'momentum'],
+    torch._ops.ops.quantized.instance_norm: [1, 2, 5, 6],
```
[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:36]: Also for `instance_norm`, is there a reason why we don't skip index 7 (eps) as well?

**Review by andrewor14 — APPROVED** [review]

File: torch/ao/quantization/fx/convert.py [hunk 3]
```diff
@@ -178,7 +184,12 @@ def remove_extra_dequantize(quantized: QuantizedGraphModule) -> QuantizedGraphMo
 
         if len(dequant_users) > 1:
             with quantized.graph.inserting_after(node):
-                unique_dq = quantized.graph.create_node("call_method", "dequantize", users[0].args, {})
+                unique_dq = quantized.graph.create_node(
+                    "call_method",
+                    "dequantize",
+                    users[0].args,
+                    {}
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/convert.py:191]: sure, will add this. technically it does not make much difference since we normalized all kwargs to args in prepare already

File: torch/ao/quantization/fx/prepare.py [hunk 4]
```diff
@@ -140,6 +142,27 @@
...
+    * inplace argument for F.relu and torch.relu since we use them
+    in pattern matching which assumes that relu only has one argument
+    """
+    for n in model.graph.nodes:
+        # remove inplace arg from relu
+        if n.op == "call_function" and n.target in (torch.nn.functional.relu, torch.relu):
+            # ignroe the inplace op since that will interfere with node
+            # matching
+            n.args = (n.args[0],)
+            n.kwargs = {}
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/prepare.py:162]: Yes, I think so, we need to otherwise inplace might be a keyword arg

File: torch/ao/quantization/fx/utils.py [hunk 5]
```diff
@@ -624,3 +627,37 @@ def create_node_from_old_node_preserve_meta(
...
     return new_node
+
+def get_all_args_as_positional_args(node: Node) -> List[Argument]:
+    """ Get a list of args and kwargs as positional args
+    """
+    all_args = list(node.args)
+    all_args.extend(list(node.kwargs.values()))
+    return all_args
+
+__all__ = [
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:638]: I see, will move

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:36]: Oh I see, got it

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37163228).

**Comment by jerryzh168** [comment]
> ## 🔗 Helpful links
> * [🧪](https://hud.pytorch.org/pr/79095)  **See artifacts and rendered test results [at hud.pytorch.org/pr/79095](https://hud.pytorch.org/pr/79095)**
> * ✖️ Python docs build was skipped
> * ✖️ C++ docs build was skipped
> * [❓](https://github.com/pytorch/pytorch/wiki/Dev-Infra-Office-Hours)Need help or want to give feedback on the CI? Visit our **[office hours](https://github.com/pytorch/pytorch/wiki/Dev-Infra-Office-Hours)**
> 
> ## ✅ No Failures (17 Pending)
> As of commit [611a584](https://github.com/pytorch/pytorch/commit/611a584a017339439fd2d9f06d4702c120d1f706) (more details [on the Dr. CI page](https://our.intern.facebook.com/intern/opensource/dr_ci/commit/167825833786582/611a584a017339439fd2d9f06d4702c120d1f706/)):
> 
> Expand to see more

actually I looked at this before landing, and this is not updated? cc @janeyx99 @malfet
